pkg/storage/consul/consul_helper.go

```diff
@@ -0,0 +1,706 @@
+			panic("unable to convert output object to pointer")
+		}
+
+		//check against len == 0, if yes what to do?
+		err = h.extractObj(storedPair, err, out, false, key)
+	}
+	return toStorageErr(err, storedPair.Key, 0)
+}
+
+func (h *consulHelper) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+	}
+	startTime := time.Now()
+
+	key = h.prefixConsulKey(key)
+	key = h.transformKeyName(key)
+	kv, _, err := h.consulKv.Get(key, nil)
+	metrics.RecordConsulRequestLatency("get", getTypeName(out), startTime)
+	if err != nil {
+		return toStorageErr(err, key, 0)
+	}
+	err = h.extractObj(kv, err, out, ignoreNotFound, key)
+	return toStorageErr(err, key, 0)
+}
+
+func (h *consulHelper) Delete(ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions) error {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+	}
+
+	key = h.prefixConsulKey(key)
+	key = h.transformKeyName(key)
+
+	v, err := conversion.EnforcePtr(out)
+	if err != nil {
+		panic("unable to convert output object to pointer: " + err.Error())
+	}
+
+	if preconditions == nil {
+		kv, _, err := h.consulKv.Get(key, &consulapi.QueryOptions{})
+		if err != nil {
+			return toStorageErr(err, key, 0)
+		}
+		if kv == nil {
+			return toStorageErr(NewNotFoundError(), key, 0)
+		}
+
+		err = h.extractObj(kv, err, out, false, key)
+		if err != nil {
+			return toStorageErr(NewInvalidObjectError(err.Error()), key, 0)
+		}
+
+		startTime := time.Now()
+		_, err = h.consulKv.Delete(key, nil)
+		metrics.RecordConsulRequestLatency("delete", getTypeName(out), startTime)
+
+		return toStorageErr(err, key, 0)
+	}
+
+	// kv declared outside of the spin-loop so that we can decode subsequent successful Gets
+	// in the event that another client deletes our key before we do.. this value is possibly
+	// lacking certified freshness
+	var kvPrev *consulapi.KVPair
+	var succeeded bool
+	// TODO: perhaps a timeout or spincount would be wise here
+	for {
+		// empty QueryOptions is explicitly setting AllowStale to false
+		startTime := time.Now()
+		kv, _, err := h.consulKv.Get(key, &consulapi.QueryOptions{})
+		metrics.RecordConsulRequestLatency("get", getTypeName(out), startTime)
+		if err != nil {
+			return toStorageErr(err, key, 0)
+		}
+		if kv == nil {
+			break
+		}
+		kvPrev = kv
+
+		obj := reflect.New(v.Type()).Interface().(runtime.Object)
+		h.extractObj(kv, err, obj, false, key)
+		if err = checkPreconditions(preconditions, obj); err != nil {
+			return toStorageErr(err, kv.Key, 0)
+		}
+
+		startTime = time.Now()
+		succeeded, _, err = h.consulKv.DeleteCAS(kv, nil)
+		metrics.RecordConsulRequestLatency("delete", getTypeName(out), startTime)
+		if err != nil {
+			return toStorageErr(err, key, 0)
+		}
+		if succeeded {
+			h.archiveEvent(key, reflect.ValueOf(obj).Type().Elem(), watch.Deleted)
+			break
+		}
+		glog.Infof("delection of %s failed because of a conflict, going to retry", key)
```

Comment:
Well, it's not needed explicitly. Tried to stay close to the [etcd implementation](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/etcd_helper.go#L191)